### PR TITLE
feat: Upgrade preview-link code sample to 3p-resources

### PR DESCRIPTION
**Pull request by PierrickVoulet**
* Remove people preview link
* Add case 3P resource creation from @ menu

### CODE REVIEW

File: apps-script/3p-resources/3p-resources.gs [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+/**
+* Extracts the URL parameters from the given URL.
+*
+* @param {!string} url The URL to parse.
+* @return {!Map} A map with the extracted URL parameters.
+*/
+function parseQuery(url) {
+  var query = url.split("?")[1];
```
Comment (vinay-google) [inline apps-script/3p-resources/3p-resources.gs:55]: `const`

File: apps-script/3p-resources/3p-resources.gs [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+  }
+}
+
+/**
+* Extracts the URL parameters from the given URL.
+*
+* @param {!string} url The URL to parse.
+* @return {!Map} A map with the extracted URL parameters.
+*/
+function parseQuery(url) {
```
Comment (vinay-google) [inline apps-script/3p-resources/3p-resources.gs:54]: Can we use something like https://developer.mozilla.org/en-US/docs/Web/API/URL to parse the URL?

File: apps-script/3p-resources/3p-resources.gs [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+ * Produces a support case creation form card.
+ * 
+ * @param {!Object} event The event object.
+ * @param {!Object=} errors An optional map of per-field error messages.
+ * @param {boolean} isUpdate Whether to return the form as an update card navigation.
+ * @return {!Card|!ActionResponse} The resulting card or action response.
+ */
+function createCaseInputCard(event, errors, isUpdate) {
+
+  const cardHeader1 = CardService.newCardHeader()
```
Comment (vinay-google) [inline apps-script/3p-resources/3p-resources.gs:90]: I only see one `cardHeader`. Is there a reason we numbered these?

File: apps-script/3p-resources/3p-resources.gs [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    impact: !!event.formInput.impact,
+  };
+
+  const errors = validateFormInputs(caseDetails);
+  if (Object.keys(errors).length > 0) {
+    return createCaseInputCard(event, errors, /* isUpdate= */ true);
+  } else {
+    const title = `Case ${caseDetails.name}`;
+    // Adds the case details as parameters to the generated link URL.
+    const url = 'https://example.com/support/cases/?' + Object.entries(caseDetails).flatMap(([k, v]) => Array.isArray(v) ? v.map(e => `${k}=${encodeURIComponent(e)}`) : `${k}=${encodeURIComponent(v)}`).join("&");
```
Comment (vinay-google) [inline apps-script/3p-resources/3p-resources.gs:191]: Should we simplify this for readability?

File: apps-script/3p-resources/3p-resources.gs [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+/**
+ * Validates case creation form input values.
+ * 
+ * @param {!Object} caseDetails The values of each form input submitted by the user.
+ * @return {!Object} A map from field name to error message. An empty object
+ *     represents a valid form submission.
+ */
+function validateFormInputs(caseDetails) {
+  const errors = {};
+  if (caseDetails.name === undefined) {
```
Comment (vinay-google) [inline apps-script/3p-resources/3p-resources.gs:208]: `if (!caseDetails.name) {`

File: apps-script/3p-resources/3p-resources.gs [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    errors.name = 'You must provide a name';
+  }
+  if (caseDetails.description === undefined) {
+    errors.description = 'You must provide a description';
+  }
+  if (caseDetails.priority === undefined) {
+    errors.priority = 'You must provide a priority';
+  }
+  if (caseDetails.impact && !(['P0', 'P1']).includes(caseDetails.priority)) {
```
Comment (vinay-google) [inline apps-script/3p-resources/3p-resources.gs:217]: Should we simplify this to:

```javascript
if (caseDetails.impact && (caseDetails.priority !== 'P0' || caseDetails.priority !== 'P1')) {
```

File: java/3p-resources/src/main/java/Create3pResources.java [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+   * @param request  An HTTP request context.
+   * @param response An HTTP response context.
+   */
+  @Override
+  public void service(HttpRequest request, HttpResponse response) throws Exception {
+    JsonObject event = gson.fromJson(request.getReader(), JsonObject.class);
+    JsonObject parameters = event.getAsJsonObject("commonEventObject").getAsJsonObject("parameters");
+    if (parameters != null && parameters.has("submitCaseCreationForm") && parameters.get("submitCaseCreationForm").getAsBoolean()) {
+      response.getWriter().write(gson.toJson(submitCaseCreationForm(event)));
+      return;
```
Comment (vinay-google) [inline java/3p-resources/src/main/java/Create3pResources.java:47]: We can remove both of the `return`s

File: java/3p-resources/src/main/java/Create3pResources.java [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+  /**
+   * Produces a support case creation form.
+   * 
+   * @param event The event object.
+   * @param errors A map of per-field error messages.
+   * @param isUpdate Whether to return the form as an update card navigation.
+   * @return The resulting card or action response.
+   */
+  JsonObject createCaseInputCard(JsonObject event, Map<String, String> errors, boolean isUpdate) {
+
```
Comment (vinay-google) [inline java/3p-resources/src/main/java/Create3pResources.java:65]: We can remove the extra space

File: java/3p-resources/src/main/java/Create3pResources.java [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+    JsonObject cardHeader1 = new JsonObject();
```
Comment (vinay-google) [inline java/3p-resources/src/main/java/Create3pResources.java:66]: Do we have another `cardHeader`/

File: java/3p-resources/src/main/java/Create3pResources.java [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+
+  /**
+   * Returns a submit form response that inserts a link into the document.
+   * 
+   * @param title The title of the link to insert.
+   * @param url The URL of the link to insert.
+   * @return The resulting submit form response.
+   */
+  JsonObject createLinkRenderAction(String title, String url) {
+    JsonObject link1 = new JsonObject();
```
Comment (vinay-google) [inline java/3p-resources/src/main/java/Create3pResources.java:313]: Should we number these if there's only of these?

**Review by vinay-google — APPROVED** [review]
LGTM! Left few nits

File: apps-script/3p-resources/3p-resources.gs (hunk 2 above)
Comment (PierrickVoulet) [reply apps-script/3p-resources/3p-resources.gs:54]: `URL` is not available from Apps Script so I need to define custom utility functions instead.

**Pull request merged**